src/Npgsql/TypeMapping/BuiltInTypeHandlerResolver.cs

```diff
@@ -338,90 +354,179 @@ internal BuiltInTypeHandlerResolver(NpgsqlConnector connector)
+                { typeof(NpgsqlCircle),  "circle" },
+                { typeof(NpgsqlLine),    "line" },
+                { typeof(NpgsqlLSeg),    "lseg" },
+                { typeof(NpgsqlPath),    "path" },
+                { typeof(NpgsqlPoint),   "point" },
+                { typeof(NpgsqlPolygon), "polygon" },
 
-            // Misc types
-            { typeof(bool), "boolean" },
-            { typeof(byte[]), "bytea" },
-            { typeof(ArraySegment<byte>), "bytea" },
+                // Misc types
+                { typeof(bool),                 "boolean" },
+                { typeof(byte[]),               "bytea" },
+                { typeof(ArraySegment<byte>),   "bytea" },
 #if !NETSTANDARD2_0
-            { typeof(ReadOnlyMemory<byte>), "bytea" },
-            { typeof(Memory<byte>), "bytea" },
+                { typeof(ReadOnlyMemory<byte>), "bytea" },
+                { typeof(Memory<byte>),         "bytea" },
 #endif
-            { typeof(Guid), "uuid" },
-            { typeof(BitArray), "bit varying" },
-            { typeof(BitVector32), "bit varying" },
-            { typeof(Dictionary<string, string>), "hstore" },
+                { typeof(Guid),                                "uuid" },
+                { typeof(BitArray),                            "bit varying" },
+                { typeof(BitVector32),                         "bit varying" },
+                { typeof(Dictionary<string, string>),          "hstore" },
 #if !NETSTANDARD2_0 && !NETSTANDARD2_1
-            { typeof(ImmutableDictionary<string, string>), "hstore" },
+                { typeof(ImmutableDictionary<string, string>), "hstore" },
 #endif
 
-            // Internal types
-            { typeof(NpgsqlLogSequenceNumber), "pg_lsn" },
-            { typeof(NpgsqlTid), "tid" },
-            { typeof(DBNull), "unknown" }
-        };
+                // Internal types
+                { typeof(NpgsqlLogSequenceNumber), "pg_lsn" },
+                { typeof(NpgsqlTid),               "tid" },
+                { typeof(DBNull),                  "unknown" }
+            };
+
+            if (LegacyTimestampBehavior)
+                ClrTypeToDataTypeNameTable[typeof(DateTime)] = "timestamp without time zone";
+        }
+
+        public override NpgsqlTypeHandler? ResolveValueDependentValue(object value)
+        {
+            // In LegacyTimestampBehavior, DateTime isn't value-dependent, and handled above in ClrTypeToDataTypeNameTable like other types
+            if (LegacyTimestampBehavior)
+                return null;
+
+            return value switch
+            {
+                DateTime dateTime => dateTime.Kind == DateTimeKind.Utc ? _timestampTzHandler : _timestampHandler,
+
+                // For arrays/lists, return timestamp or timestamptz based on the kind of the first DateTime; if the user attempts to
+                // mix incompatible Kinds, that will fail during validation. For empty arrays it doesn't matter.
```

Comment:
Validation of the subhandler? I wouldn't say those validation errors really point users to the source of their issue when they're unknowingly mixing DT.Kinds.

```
    $"Cannot write DateTime with Kind={value.Kind} to PostgreSQL type 'timestamp with time zone', only UTC is supported. " +
                    "See the Npgsql.EnableLegacyTimestampBehavior AppContext switch to enable legacy behavior.");
```

Or is there another mechanism I've overlooked?

Reply:
No, that's it. We can do a special DateTimeArrayHandler which adds validation for a better error message, but the above error actually seemed somewhat decent to me... Should we wait for confused users and then see?

Reply:
We could also just mention arrays/ranges here?

Reply:
Sure, why not